### PR TITLE
UI Press Book content type first draft

### LINKED ISSUE #4328: UI Press - create Book content type
This will replace UI Press Book content type currently used on their d7 site.

General layout - customer would like something similar to an amazon book page, with the book cover on the left, text about the book in the middle, and buying options on the right. Book cover image needs to be larger than it is on current site.

Fields & their current formats, with some possible changes:
- Title - text field, required
- Subtitle - text field
- Book cover - image upload.
- Author(s) - autocomplete from Person content type
- Editor(s) - autocomplete from Person content type
- Illustrator(s) - autocomplete from Person content type
- Photographer(s) - autocomplete from Person content type
- Translator(s) - autocomplete from Person content type
- Preface author(s) - autocomplete from Person content type
- Foreword author(s) - autocomplete from Person content type
- Introduction author(s) - autocomplete from Person content type
- Afterword author(s) - autocomplete from Person content type
- Essay by - autocomplete from Person content type
- ~As told to - autocomplete from Person content type~ checking if needed
- Awards - text area
- Publication date - should display month & year
- Pages - text field
- Art - text field
- Trim size - text field
- Season - taxonomy (new vocab)
- Edition - radio button selection, required. Options: 1st, 2nd, 3rd, 4th, 5th.
- Hardcover retail price - decimal field
- Hardcover sale price -  decimal field
- Hardcover sale code - text field
- Hardcover sale start date - mm/dd/yyyy
- Hardcover sale expiry date - mm/dd/yyyy
- Hardcover ISBN-13 - string
- Paperback retail price - decimal field
- Paperback sale price -  decimal field
- Paperback sale code - text field
- Paperback sale start date - mm/dd/yyyy
- Paperback sale expiry date - mm/dd/yyyy
- Paperback ISBN-13 - string
- eBook retail price - decimal field
- eBook sale price -  decimal field
- eBook sale code - text field
- eBook sale start date - mm/dd/yyyy
- eBook sale expiry date - mm/dd/yyyy
- eBook ISBN-13 - string
- Reviews/blurbs - text area
- Book description/summary - text area
- Contributors - text area
- Table of contents - text area
- Excerpt - text area 
- Series - term reference, unlimited. Add Series as a vocabulary.
- Subject(s) - term reference, unlimited. Add Subjects as a vocabulary.
- Related Reads - autocomplete from Book content type, would like option for 5 related reads
- File uploads - they would like a section to upload images from the book, shared example of site that had modal image gallery.

They would like a way to share the book info straight to social media, like the 'Share Title' section [halfway down this page](https://www.graywolfpress.org/books/my-name-will-grow-wide-tree). Did not commit, said we would investigate further.

Instead of listing every field for Hardcover, Paperback, and eBook, it would be better to have them as sections that can be expanded as needed. Also, sale prices should only display when a sale is active.

### LINKED ISSUE #4329: UI Press - Investigate using Person content type for creating Authors
# Motivation

UI Press needs a Person content type to handle authors, illustrators, etc that contribute to a book but aren't part of their staff. In D7 they used a separate Author content type. The different roles could be added to our existing Person Types, but since there's about a dozen it might be easier to keep them separate and create a custom Person type. 

Fields needed:
- First Name
- Last Name
- Roles/Person Types 
   - [x] Author
   - [x] Editor
   - [x] Illustrator
   - [x] Photographer
   - [x] Translator
   - [x] Preface Author
   - [x] Foreword Author
   - [x] Introduction Author
   - [x] Afterword Author
   - [x] Essay by
   - [x] Contributor 
- Biography 
- Website 
- Image 
  - Place to list photographer credit for author photo
- Social media links? - nice-to-have, existing link fields should fulfill this

Additionally, in d7 the page that displays for each person includes a view that lists & links any books they worked on and what their role was. (https://github.com/uiowa/uiowa/issues/4658)

# Proposed solution

Investigate how to set up person types (using custom person types or just tags) and what additional customizations would be necessary (if any) to use the Person content type for authors/etc. in the new UI Press site. We also need to investigate what are the potential problems with using the Person content type (e.g. the built-in person listing page needs to be turned off, what is involved in building out listing pages for authors, etc.)

### PR DESCRIPTION
**Pull request by makurj**
Resolves #4328 
Resolves #4329 

How to test:
1. blt ds  --site=uipress.uiowa.edu
2. Enable UI Press_Core module (drush @uipress.local en uipress_core)
3. drush -l uipress.uiowa.edu config:import --source ../config/sites/uipress.uiowa.edu --partial
5. drush @uipress.local cr
6. Create a book content type
7. Add some fields
8. Check sale logic by entering price and/or dates
9. If the sale price isn't added or if you're not within the sale date range, sale dates/price/code should not display. 
10. Test the opposite condition.

### CODE REVIEW

File: docroot/sites/uipress.uiowa.edu/modules/uipress_core/uipress_core.module [hunk 1]
```diff
@@ -14,11 +14,8 @@ function uipress_core_preprocess_paragraph__book_type(&$variables)
   if (!$admin_context->isAdminRoute()) {
     $paragraph = $variables['paragraph'];
     if ($paragraph->bundle() == 'book_type') {
-      if ($paragraph->field_book_sale->value == 1) {
-        $variables["content"]["field_book_sale_code"] = display_title = true;
-        unset($variables["content"]["field_book_sale_expiry_date"]);
-        unset($variables["content"]["field_book_sale_price"]);
-        unset($variables["content"]["field_book_sale_start_date"]);
+      if ($paragraph->field_book_sale->value == 0) {
+        unset($variables["content"]["field_book_sale_expiry_date"], $variables["content"]["field_book_sale_price"], $variables["content"]["field_book_sale_start_date"], $variables["content"]["field_book_sale_code"]);
```
Comment (pyrello) [inline docroot/sites/uipress.uiowa.edu/modules/uipress_core/uipress_core.module]: Can we switch these double quotes to single quotes?

Comment (dnkemper) [reply docroot/sites/uipress.uiowa.edu/modules/uipress_core/uipress_core.module]: bahaha we knew you had a preference, just couldn't remember what it was. Deal!

Comment (makurj) [reply docroot/sites/uipress.uiowa.edu/modules/uipress_core/uipress_core.module]: On it...

**Comment by joewhitsitt** [comment]
```
FILE: ...sites/uipress.uiowa.edu/modules/uipress_core/uipress_core.module
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 27 | ERROR | [x] Space before opening parenthesis of function call prohibited
```

**Comment by dnkemper** [comment]
thank you, @joewhitsitt! Fixed.

**Comment by pyrello** [comment]
I'm going to take a look at this today.

**Comment by pyrello** [comment]
This is looking really good. I made some adjustments based on my testing. Mostly to allow for creating of new terms when filling in Series, Season, and Subject(s). Also swapped out the image field to use the existing `field_image` media reference field.

The remaining thing that needs to be tackled before this is done is to handle a field that is specific to e-books. They have an extra field that other book types don't have to indicate the type of license (120 days or perpetual). I'm going to add this as an extra field to the book type paragraph bundle and then add some preprocess logic to only show it when e-book is selected as the type.

**Review by pyrello — APPROVED** [review]
This looks good. There may still be additional tweaks, but they can be handled in follow-up issues. Thanks to @makurj and @dnkemper for their work on this!

**Comment by dnkemper** [comment]
YAAASSS! Thank you, @pyrello! @makurj, step 1 complete. UP NEXT: Author view work. WOOHOOO!

**Pull request merged**